types/ramda/index.d.ts

```diff
@@ -1027,6 +1028,13 @@ declare namespace R {
          */
         memoize<T = any>(fn: (...a: any[]) => T): (...a: any[]) => T;
 
+        /**
+         * A customisable version of R.memoize. memoizeWith takes an additional function that will be applied to a given
+         * argument set and used to create the cache key under which the results of the function to be memoized will be stored.
+         * Care must be taken when implementing key generation to avoid clashes that may overwrite previous entries erroneously.
+         */
+        memoizeWith<T = any>(keyFn: (...v: any[]) => string, fn: T): T;
```

Comment:
The type of `fn` and the return value don't look right to me. Shouldn't they be the same as for the older `memoize`?

Reply:
Actually, I was considering to update a `memoize` too as current `memoize` type def saying that a returned function will accept any params. However, a type of a returned function **should** be the same as a type of `fn`. 

In another words, by using a `memoize` you will throw away all type information. By using a `memoizeWith`, type will be preserved.

Reply:
Oh right I see your point. What threw me off was the `= any` in the type parameter. With these types it's still to do this:
```
const memoized = memoizeWith(JSON.stringify, 4);
```
where typescript will say `memoized` has the type `number`.

After a quick test, if the type parameter was changed from `<T = any>` to `<T extends (...args: any[]) => any>`, only function types will be allowed, and the return type will have the same type as the `fn` parameter. Do you think that would work here?

Reply:
Well, I've actually tried to use `<T extends Function>` but Travis complained about this...
Your solution should work as it is more specific. I will do it right now.